### PR TITLE
Add Kamal by default to Rails 8

### PR DESCRIPTION
**Pull request by dhh**
Rails should have a default answer to deploying applications out of the box to cloud or bare metal. Kamal is that answer. We'll configure it automatically as far as we can, but allow an opt-out with `--skip-kamal`.

### CODE REVIEW

**Comment by dhh** [comment]
cc @djmb

File: railties/lib/rails/generators/rails/app/templates/env.erb.tt [hunk 1]
```diff
@@ -0,0 +1,5 @@
+<%%# This env template should lookup values from a password store or from ENV -%>
+# Generated by kamal envify from .env.erb
+KAMAL_REGISTRY_PASSWORD=<%%= "YOUR-KEY-FROM-PASSWORD-STORE" %>
+RAILS_MASTER_KEY=<%%= File.read("config/master.key") %>
+# GITHUB_TOKEN=<%%#= `gh config get -h github.com oauth_token`.strip %>
```
Comment (hey-leon) [inline railties/lib/rails/generators/rails/app/templates/env.erb.tt:5]: This is a neat idea but can we can rely on having the `gh` cli?

Comment (dhh) [reply railties/lib/rails/generators/rails/app/templates/env.erb.tt:5]: No, which is why it's commented out by default. So it's just a suggestion for how to do this.

**Comment by dhh** [comment]
Unrelated flaky build failures.

File: railties/lib/rails/generators/rails/app/templates/bin/kamal.tt
[the remark this answers is not in the document]
Comment (dhh) [reply railties/lib/rails/generators/rails/app/templates/bin/kamal.tt:1]: Move this to have bundler generate the binstub. We should do this for the others as well.

**Comment by dhh** [comment]
The flaky tests that keep tripping intermittently:

image

**Comment by dhh** [comment]
Closes https://github.com/rails/rails/issues/50441

**Comment by laptopmutia** [comment]
some people really work like a beast!

thanks for this

File: railties/lib/rails/generators/rails/app/templates/env.erb.tt (hunk 1 above)
Comment (hey-leon) [reply railties/lib/rails/generators/rails/app/templates/env.erb.tt:5]: 🤦‍♂️Ofcourse. Missed the comment without the highlighting.

**Comment by plattenschieber** [comment]
Should we also adapt the [base kamal deploy config](https://github.com/basecamp/kamal/blob/main/lib/kamal/cli/templates/deploy.yml) to reflect [the new default config](https://github.com/rails/rails/blob/main/railties/lib/rails/generators/rails/app/templates/config/deploy.yml.tt)? 


<details>
  <summary>The diff</summary>
  
  ```
  diff --git a/config/deploy.yml b/config/deploy.yml
index de5a4d4..2602c66 100644
--- a/config/deploy.yml
+++ b/config/deploy.yml
@@ -1,23 +1,18 @@
 # Name of your application. Used to uniquely configure containers.
-service: application
+service: my-app
 
 # Name of the container image.
-image: your-user/application
+image: user/my-app
 
 # Deploy to these servers.
 servers:
-  web:
-    - 192.168.0.1
-  # job:
-  #   hosts:
-  #     - 192.168.0.1
-  #   cmd: bin/solid_queue work
+  - 192.168.0.1
 
 # Credentials for your image host.
 registry:
   # Specify the registry server, if you're not using Docker Hub
   # server: registry.digitalocean.com / ghcr.io / ...
-  username: your-user
+  username: my-user
 
   # Always use an access token rather than real password when possible.
   password:
@@ -25,43 +20,25 @@ registry:
 
 # Inject ENV variables into containers (secrets come from .env).
 # Remember to run `kamal env push` after making changes!
-env:
-  secret:
-    - RAILS_MASTER_KEY
-  # clear:
-  #   DB_HOST: 192.168.0.2
-
-
-# Use a persistent storage volume for sqlite database files and local Active Storage files.
-# Recommended to change this to a mounted volume path that is backed up off server.
-volumes:
-  - "application_storage:/rails/storage"
-
-
-# Bridge fingerprinted assets, like JS and CSS, between versions to avoid
-# hitting 404 on in-flight requests. Combines all files from new and old
-# version inside the asset_path.
-asset_path: /rails/public/assets
+# env:
+#   clear:
+#     DB_HOST: 192.168.0.2
+#   secret:
+#     - RAILS_MASTER_KEY
 
 # Use a different ssh user than root
 # ssh:
 #   user: app
 
-# Configure builder setup (defaults to multi-arch images).
+# Configure builder setup.
 # builder:
-#   # Build same-arch image locally (use for x86->x86)
-#   multiarch: false
-#
-#   # Build diff-arch image via remote server
-#   remote:
-#     arch: amd64
-#     host: ssh://app@192.168.0.1
-#
 #   args:
-#     RUBY_VERSION: 3.3.1
+#     RUBY_VERSION: 3.2.0
 #   secrets:
 #     - GITHUB_TOKEN
-#     - RAILS_MASTER_KEY
+#   remote:
+#     arch: amd64
+#     host: ssh://app@192.168.0.1
 
 # Use accessory services (secrets come from .env).
 # accessories:
@@ -85,3 +62,40 @@ asset_path: /rails/public/assets
 #     port: 6379
 #     directories:
 #       - data:/data
+
+# Configure custom arguments for Traefik. Be sure to reboot traefik when you modify it.
+# traefik:
+#   args:
+#     accesslog: true
+#     accesslog.format: json
+
+# Configure a custom healthcheck (default is /up on port 3000)
+# healthcheck:
+#   path: /healthz
+#   port: 4000
+
+# Bridge fingerprinted assets, like JS and CSS, between versions to avoid
+# hitting 404 on in-flight requests. Combines all files from new and old
+# version inside the asset_path.
+#
+# If your app is using the Sprockets gem, ensure it sets `config.assets.manifest`.
+# See https://github.com/basecamp/kamal/issues/626 for details
+#
+# asset_path: /rails/public/assets
+
+# Configure rolling deploys by setting a wait time between batches of restarts.
+# boot:
+#   limit: 10 # Can also specify as a percentage of total hosts, such as "25%"
+#   wait: 2
+
+# Configure the role used to determine the primary_host. This host takes
+# deploy locks, runs health checks during the deploy, and follow logs, etc.
+#
+# Caution: there's no support for role renaming yet, so be careful to cleanup
+#          the previous role on the deployed hosts.
+# primary_role: web
+
+# Controls if we abort when see a role with no hosts. Disabling this may be
+# useful for more complex deploy configurations.
+#
+# allow_empty_roles: false

  ```
</details>

**Comment by dhh** [comment]
No, the base Kamal can't be Rails specific. Lots of Kamal users in other environments than Rails.

**Pull request merged**